docs/HRANA_2_SPEC.md

```diff
@@ -91,6 +93,29 @@ the response.
 It is not an error if the client attempts to close a SQL text id that is not
 used.
 
+### Execute a sequence of SQL statements
+
+```typescript
+type SequenceReq = {
+    "type": "sequence",
+    "stream_id": int32,
+    "sql"?: string | null,
+    "sql_id"?: int32 | null,
+}
+
+type SequenceResp = {
+    "type": "sequence",
+}
+```
+
+The `sequence` request executes a sequence of SQL statements separated by
+semicolons on the stream given by `stream_id`. `sql` or `sql_id` specify the SQL
+text; exactly one of these fields must be specified.
+
+Any rows returned by the statements are ignored. If any statement fails, the
```

Comment:
Why are we not returning any rows? if not for all select statements then at least for the last one?

Reply:
The `sequence` request is designed to implement client APIs like:
- [`Cursor.executescript()`](https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.executescript) in Python
- [`Database.exec()`](https://github.com/TryGhost/node-sqlite3/wiki/API#execsql--callback) in Node.js `sqlite3`
- [`Connection.execute_batch()`](https://docs.rs/rusqlite/latest/rusqlite/struct.Connection.html#method.execute_batch) in Rust `rusqlite`

These methods don't return any rows, and some of them even report an error if some of the statements do! Therefore, to keep the protocol and the implementation simple, no results are returned in Hrana 2. We can always extend the protocol to optionally return the results later, if it turns out that we need it.

Reply:
execute_batch could return results. I'm asking because turso shell will support executing batch of statements separated with ; and it is useful to return values. For example BEGIN; INSERT INTO t ...; SELECT FROM t; COMMIT

Reply:
Won't the shell need to parse the semicolons anyway, to detect when the input is finished?

Reply:
it may parse it but it will have to send all statements as a single batch in the end anyway

Reply:
To send a batch of statements, the shell should use the `batch` request.

Reply:
I like @haaawk's idea to allow for `want_rows=true` in the request and receive something similar to `BatchResult`, but we'd need another array with the statements, such as it would be implemented as a `batch` where every line depends on the previous not failing.

```typescript
type SequenceResult = BatchResult & { "stmts": string[] };
```

Reply:
batch is not providing the right semantic. `BEGIN;INSERT...;SELECT...;COMMIT` run in batch would not be atomic would it?
I could also want `BEGIN;...;COMMIT;BEGIN;...;COMMIT`. It would definitely not work as expected with batch, will it @honzasp ?

Reply:
What would be the use for this?